File: src/main/cpp/src/SROTestDataType.cpp
#ifndef ERSAP_DEMO_DATA_SRO_HPP_
#define ERSAP_DEMO_DATA_SRO_HPP_

#include <ersap/engine_data_type.hpp>
#include <ersap/any.hpp>
#include <vector>
#include <cstdint>
#include <memory>
#include <stdexcept>
#include <cstring>
#include <SROTestDataType.hpp>  // Assume this includes RocTimeFrameBank and FADCHit definitions

namespace ersap {
namespace demo {

using byte_t = std::uint8_t;
using bytes_t = std::vector<byte_t>;

class ByteBuffer {
public:
    ByteBuffer(size_t size) { data_.reserve(size); }

    void put(std::int32_t v) {
        data_.insert(data_.end(), (byte_t*) &v, (byte_t*) &v + sizeof(v));
    }

    void put(std::int64_t v) {
        data_.insert(data_.end(), (byte_t*) &v, (byte_t*) &v + sizeof(v));
    }

    template<typename T>
    void putRange(const T& v) {
        data_.insert(data_.end(), v.begin(), v.end());
    }

    const bytes_t& data() const { return data_; }

public:
    bytes_t data_;
};

class SROSerializer : public ersap::Serializer {
public:
    std::vector<std::uint8_t> write(const ersap::any& data) const override {
        const auto& sro = ersap::any_cast<const std::vector<std::vector<RocTimeFrameBank>>&>(data);

        ByteBuffer buffer(8192); // starting with an arbitrary size
        buffer.put((int32_t)sro.size());

        for (const auto& sublist : sro) {
            buffer.put((int32_t)sublist.size());
            for (const auto& frame : sublist) {
                buffer.put((int32_t)frame.getRocID());
                buffer.put((int32_t)frame.getFrameNumber());
                buffer.put((int64_t)frame.getTimeStamp());

                const auto& hits = frame.getHits();
                buffer.put((int32_t)hits.size());
                for (const auto& hit : hits) {
                    buffer.put((int32_t)hit.crate());
                    buffer.put((int32_t)hit.slot());
                    buffer.put((int32_t)hit.channel());
                    buffer.put((int32_t)hit.charge());
                    buffer.put((int64_t)hit.time());
                }
            }
        }

        return buffer.data_;
    }

    ersap::any read(const std::vector<std::uint8_t>& buffer) const override {
        size_t i = 0;
        int32_t outer_size = to_int(buffer, i); i += 4;
        std::vector<std::vector<RocTimeFrameBank>> sro;
        sro.reserve(outer_size);

        for (int32_t o = 0; o < outer_size; ++o) {
            int32_t inner_size = to_int(buffer, i); i += 4;
            std::vector<RocTimeFrameBank> sublist;
            sublist.reserve(inner_size);

            for (int32_t f = 0; f < inner_size; ++f) {
                RocTimeFrameBank frame;
                frame.setRocID(to_int(buffer, i)); i += 4;
                frame.setFrameNumber(to_int(buffer, i)); i += 4;
                frame.setTimeStamp(to_long(buffer, i)); i += 8;

                int32_t hit_count = to_int(buffer, i); i += 4;
                for (int32_t h = 0; h < hit_count; ++h) {
                    int crate = to_int(buffer, i); i += 4;
                    int slot = to_int(buffer, i); i += 4;
                    int channel = to_int(buffer, i); i += 4;
                    int charge = to_int(buffer, i); i += 4;
                    long time = to_long(buffer, i); i += 8;
                    frame.addHit(FADCHit(crate, slot, channel, charge, time));
                }
                sublist.push_back(frame);
            }
            sro.push_back(sublist);
        }

        return ersap::any{sro};
    }

private:
    int32_t to_int(const bytes_t& b, size_t offset) const {
        return (b[offset+3] << 24) | (b[offset+2] << 16) | (b[offset+1] << 8) | (b[offset]);
    }

    int64_t to_long(const bytes_t& b, size_t offset) const {
        int64_t val = 0;
        for (int j = 7; j >= 0; --j) {
            val = (val << 8) | b[offset + j];
        }
        return val;
    }
};

const ersap::EngineDataType SRO_TYPE{"binary/sro-data", std::make_unique<SROSerializer>()};

} // namespace demo
} // namespace ersap

#endif

File: src/main/java/org/jlab/ersap/actor/datatypes/SROTestDataType.java
package org.jlab.ersap.actor.datatypes;

import org.jlab.epsci.ersap.base.error.ErsapException;
import org.jlab.epsci.ersap.engine.EngineDataType;
import org.jlab.epsci.ersap.engine.ErsapSerializer;
import org.jlab.ersap.actor.coda.proc.RocTimeFrameBank;
import org.jlab.ersap.actor.coda.proc.EtEvent;
import org.jlab.ersap.actor.coda.proc.FADCHit;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;
import java.util.ArrayList;

/**
 * A ERSAP engine data-type for a List<List<RocTimeFrameBank>>
 */
public final class SROTestDataType extends EngineDataType {

    private static final String MIME_TYPE = "binary/sro-data";

    public static final SROTestDataType INSTANCE = new SROTestDataType();


    private SROTestDataType() {
        super(MIME_TYPE, new ErsapSerializer() {

            @Override
            public ByteBuffer write(Object data) throws ErsapException {
                List<List<RocTimeFrameBank>> sroData;
                System.out.println("DDD ===> "+ data.getClass().getName());
                // Handle both EtEvent and direct List<List<RocTimeFrameBank>>
                if (data instanceof EtEvent) {
                    EtEvent etEvent = (EtEvent) data;
                    sroData = etEvent.getTimeFrames();
                } else {
                    sroData = (List<List<RocTimeFrameBank>>) data;
                }
                
                try {
                    // Use the same binary format as C++ for compatibility
                    ByteArrayOutputStream bos = new ByteArrayOutputStream();
                    DataOutputStream out = new DataOutputStream(bos);
                    
                    // Write outer list size
                    System.out.println("DDD-Java: Writing outer size: " + sroData.size());
                    out.writeInt(sroData.size());
                    
                    for (List<RocTimeFrameBank> sublist : sroData) {
                        // Write inner list size
                        System.out.println("DDD-Java: Writing inner size: " + sublist.size());
                        out.writeInt(sublist.size());
                        
                        for (RocTimeFrameBank frame : sublist) {
                            // Write frame data
                            out.writeInt(frame.getRocID());
                            out.writeInt(frame.getFrameNumber());
                            out.writeLong(frame.getTimeStamp());
                            
                            // Write hits
                            List<FADCHit> hits = frame.getHits();
                            out.writeInt(hits.size());
                            
                            for (FADCHit hit : hits) {
                                out.writeInt(hit.crate());
                                out.writeInt(hit.slot());
                                out.writeInt(hit.channel());
                                out.writeInt(hit.charge());
                                out.writeLong(hit.time());
                            }
                        }
                    }
                    
                    out.flush();
                    byte[] bytes = bos.toByteArray();
                    System.out.println("DDD-Java: Total bytes written: " + bytes.length);
                    ByteBuffer buffer = ByteBuffer.allocate(bytes.length);
                    buffer.put(bytes);
                    buffer.flip();
                    return buffer;
                } catch (IOException e) {
                    throw new ErsapException("Failed to serialize SRO data", e);
                }
            }

            @Override
            public Object read(ByteBuffer buffer) throws ErsapException {
                try {
                    byte[] bytes = new byte[buffer.remaining()];
                    buffer.get(bytes);
                    ByteArrayInputStream bis = new ByteArrayInputStream(bytes);
                    DataInputStream in = new DataInputStream(bis);
                    
                    // Read outer list size
                    int outerSize = in.readInt();
                    List<List<RocTimeFrameBank>> timeFrames = new ArrayList<>();
                    
                    for (int o = 0; o < outerSize; ++o) {
                        // Read inner list size
                        int innerSize = in.readInt();
                        List<RocTimeFrameBank> sublist = new ArrayList<>();
                        
                        for (int f = 0; f < innerSize; ++f) {
                            RocTimeFrameBank frame = new RocTimeFrameBank();
                            frame.setRocID(in.readInt());
                            frame.setFrameNumber(in.readInt());
                            frame.setTimeStamp(in.readLong());
                            
                            // Read hits
                            int hitCount = in.readInt();
                            for (int h = 0; h < hitCount; ++h) {
                                int crate = in.readInt();
                                int slot = in.readInt();
                                int channel = in.readInt();
                                int charge = in.readInt();
                                long time = in.readLong();
                                frame.addHit(new FADCHit(crate, slot, channel, charge, time));
                            }
                            sublist.add(frame);
                        }
                        timeFrames.add(sublist);
                    }
                    
                    // Create an EtEvent to wrap the timeFrames for consistency
                    EtEvent etEvent = new EtEvent();
                    for (List<RocTimeFrameBank> timeFrame : timeFrames) {
                        etEvent.addTimeFrame(timeFrame);
                    }
                    return etEvent;
                } catch (IOException e) {
                    throw new ErsapException("Failed to deserialize SRO data", e);
                }
            }
        });

    }
}
